text/0044-rent-based-registration.md

```diff
@@ -0,0 +1,199 @@
+
+For registering using the new rental system we will have to make modifications to the `paras-registrar` pallet. We should expose two new extrinsics for this:
+```rust
+mod pallet {
+	// -- snip --
+
+	pub fn register_rental(
+		origin: OriginFor<T>,
+		id: ParaId,
+		genesis_head: HeadData,
+		validation_code: ValidationCode,
+	) -> DispatchResult { /* ... */ }
+
+	pub fn pay_rent(origin: OriginFor<T>, id: ParaId) -> DispatchResult {
+		/* ... */ 
+	}
+}
+```
+
+A call to `register_rental` will require the reservation of only a percentage of the deposit that would otherwise be required to register the validation code when using the regular model.
+As described later in the *Quick para re-registering* section below, we will also store the code hash of each parachain to enable faster re-registration after a parachain has been pruned. For this reason the total initial deposit amount is increased to account for that.
+```rust
+// The logic for calculating the initial deposit for parachain registered with the 
+// new rent-based model:
+
+let validation_code_deposit = per_byte_fee.saturating_mul((validation_code.0.len() as u32).into());
+
+let head_deposit = per_byte_fee.saturating_mul((genesis_head.0.len() as u32).into())
+let hash_deposit = per_byte_fee.saturating_mul(HASH_SIZE);
+
+let deposit = T::InitialRentDeposit::get().mul_ceil(validation_code_deposit)
+	.saturating_add(T::ParaDeposit::get())
+	.saturating_add(head_deposit)
+	.saturating_add(hash_deposit)
+```
+Once the `ParaId` is reserved and the validation code is registered the rent must be periodically paid to ensure the on-demand parachain doesn't get removed from the state. The `pay_rent` extrinsic should be callable by anyone, removing the need for the parachain to depend on the parachain manager for rent payments.
+
+### On-demand parachain pruning
+
+If the rent is not paid, anyone has the option to prune the on-demand parachain and claim a portion of the initial deposit reserved for storing the validation code. Pruning only removes the validation code, while the head data and validation code hash are retained. The validation code hash is stored to allow anyone to register it again as well as to enable quicker re-registration by skipping the pre-checking process.
+
+The moment the rent is no longer paid, the parachain won't be able to purchase on-demand access, meaning no new blocks are allowed. This stage is called the "hibernation" stage, during which all the parachain-related data is still stored on-chain, but new blocks are not permitted. The reason for this is to ensure that the validation code is available in case it is needed in the dispute or approval checking subsystems. Waiting for one entire session will be enough to ensure it is safe to deregister the parachain.
```

Comment:
It is a soft form of deregistration though. ParaId is still claimed, only the PVF is pruned.

Reply:
Yes, I think it can be called so. The main reason for this is to allow anyone to re-register the parachain under the same `ParaId` after it has been pruned.

Of course, there is always the option to perform a complete deregistration of a parachain by calling the `deregister` extrinsic.